Bluetooth: controller: split: Port h/w privacy feature

cvinayak opened the pull request:
Ported h/w based LL privacy to ULL/LLL split architecture.

Fixes #14675.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Port h/w privacy feature

Ported h/w based LL privacy to ULL/LLL split architecture.

Fixes #14675.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Revert "Bluetooth: controller: Disable controller privacy feature"

This reverts commit c629ea8a64c647a1ff1a31088d8fe15970254980.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented on subsys/bluetooth/controller/ll_sw/lll_filter.h:
```diff
@@ -7,11 +7,21 @@
 
-struct ll_filter *ctrl_filter_get(bool whitelist);
-void ll_adv_scan_state_cb(u8_t bm);
+extern bt_addr_t *ull_filter_lll_lrpa_get(u8_t rl_idx);
```
why can't these functions be just named `lll_*()` without the `ull_filter_` prefix?

pskhansen replied:
I think the notation is that this is an interface provided by ull_filter but is executed in lll context

pskhansen reviewed (approved):
Looks fine.

cvinayak commented:
Initiator tests failing, until fixed, I will keep DNM label on this PR. Otherwise 30 of 39 Privacy related tests are passing.

Suspecting a initiator bug, and am investigating it.

joerchan committed:
Bluetooth: controller: Fix initialization of rl_idx and rpa_gen

The privacy state variables rl_idx and rpa_gen was not initialized
properly, which results in connection failing to be established.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: controller: Fix pdu being overwritten while in use

The node_rx buffer for the connect ind buffer is being reused to send
connection complete event, make sure all data is backed up before
reusing the buffer.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

joerchan committed:
Bluetooth: controller: Copy the connect_ind into node_rx

The node_rx buffer does not contain the data from the connect ind, so
copy it over.
Also back up the adv pdu chan_sel bit so that it can be used to properly
select the correct channel selection algorithm

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

joerchan commented:
These tests are currently failing in the EBQ.
/LL/CON/INI/BV-11-C: Failure
/LL/CON/INI/BV-12-C: Failure
/LL/CON/INI/BV-24-C: Failure

The reason is an assert in ull.c:417
This seems to happen when the master receives the first control PDU (no data yet) on the link
The control pdu is LL_SLAVE_FEATURE_REQ

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_master.c:
```diff
@@ -434,13 +434,21 @@ void ull_master_setup(memq_link_t *link, struct node_rx_hdr *rx,
 	chan_sel = pdu->chan_sel;
 
+#if defined(CONFIG_BT_CTLR_PRIVACY)
+	u8_t init_addr[BDADDR_SIZE];
```
General comment, what is the convention when it comes to variable declaration, should it be at the beginning of the locality of reference or can it be as-is close to its first use?

carlescufi replied:
It should be at the beginning of the function if possible, otherwise wherever it makes sense with #fidef statements

joerchan committed:
Bluetooth: controller: Fix compilation without broadcaster or observer

Fix compilation issues in central_hr or peripheral_hr sample when using
split LL and privacy is enabled.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mtpr-ot commented:
> These tests are currently failing in the EBQ.
> /LL/CON/INI/BV-11-C: Failure
> /LL/CON/INI/BV-12-C: Failure
> /LL/CON/INI/BV-24-C: Failure
> 
> The reason is an assert in ull.c:417
> This seems to happen when the master receives the first control PDU (no data yet) on the link
> The control pdu is LL_SLAVE_FEATURE_REQ

I have tried to run the three test cases on BabbleSim EDTT. Unfortunately I'm not able to reproduce the assertion. The results are:
**Old stack (PR-16037):**
/LL/CON/INI/BV-11-C: Failure
/LL/CON/INI/BV-12-C: Pass
/LL/CON/INI/BV-24-C: Pass
**New stack (PR-16037):**
/LL/CON/INI/BV-11-C: Failure
/LL/CON/INI/BV-12-C: Pass
/LL/CON/INI/BV-24-C: Pass

joerchan commented:
```+CONFIG_CLOCK_CONTROL_NRF_K32SRC_150PPM=y
+
+CONFIG_BT_CTLR_TX_BUFFER_SIZE=251
+CONFIG_BT_RX_BUF_LEN=258
+CONFIG_BT_CTLR_DATA_LENGTH_MAX=251
+CONFIG_BT_CTLR_ADVANCED_FEATURES=y
+CONFIG_BT_CTLR_FAST_ENC=y
+CONFIG_BT_CTLR_CONN_RSSI=y
```
Using those configurations in the hci_uart sample, the assert goes away.
Something is wrong in memory handling, but I can't find anything that would have been introduced by this PR. So I think we should go ahead.

The test /LL/CON/INI/BV-24-C: is failing for both legacy and split LL in EBQ. 
Since this is a failure for both, and the test is under waiver and needs to be clarified. We will skip this test failure for now.
The test specification and the EBQ test case is also not expecting the same result.

aescolar commented:
@joerchan  Do you want to remove the DNM tag and prefix in the title?

cvinayak commented:
@joerchan there is spurious "Squash! .. "message in one of the commit messages. Please fix it.

joerchan commented:
> @joerchan there is spurious "Squash! .. "message in one of the commit messages. Please fix it.

Thanks, fixed.

joerchan commented:
> @joerchan Do you want to remove the DNM & RFC tags and prefix in the title?

Yes, done.

After rebasing the PR onto master, the assert is now gone.
The commit that fixes it is this: 58e9ac6811c300a2fdcefd23dd6d6a87c9e20829 (reverting it brings the assert back).

mtpr-ot reviewed (approved):
No difference between legacy- and split implementation qualification test results on BabbleSim/EDTT. I think we should approve despite the one remaining EBQ test failing.

aescolar commented:
@carlescufi There is no complaints from here, so if you and @cvinayak are ok with it, feel free to press the button.

cvinayak commented:
ull_slave is missing privacy code, leading to pairing failure due to missing local_rpa value in connection complete event. putting back into DNM.

cvinayak commented:
@carlescufi This PR is ready to be merged.
